### PR TITLE
New service framework

### PR DESCRIPTION
**Pull request by yuriks**
From the main commit's description:

> The old "Interface" class had a few problems such as using free
> functions (Which didn't allow you to write the service handler as if it
> were a regular class.) which weren't very extensible. (Only received one
> parameter with a pointer to the Interface object.)
> 
> The new ServiceFramework aims to solve these problems by working with
> member functions and passing a generic context struct as parameter. This
> struct can be extended in the future without having to update all
> existing service implementations.

Please hold off on actually migrating stuff to this new framework for now. I have some other API changes planned/half-implemented which, while backwards compatible, should save a lot of work when writing handlers, so it's probably better to wait for them.

---
This change is [Reviewable](https://reviewable.io/reviews/citra-emu/citra/2756)

### CODE REVIEW

File: src/core/hle/service/service.h [hunk 1]
```diff
@@ -101,6 +109,147 @@ class Interface : public Kernel::SessionRequestHandler {
...
+    struct FunctionInfoBase {
+        u32 expected_header;
+        HandlerFnP<ServiceFrameworkBase> handler_callback;
+        const char* name;
+    };
+
+    using InvokerFn = void(ServiceFrameworkBase* object, HandlerFnP<ServiceFrameworkBase> member,
+                           Kernel::HLERequestContext& ctx);
+
+    ServiceFrameworkBase(const char* service_name, unsigned int max_sessions,
```
Comment (lioncash) [inline src/core/hle/service/service.h]: Should these this and the class member be using `u32` instead to be consistent with the other functions?

File: src/core/hle/service/service.h [hunk 2]
```diff
@@ -101,6 +109,147 @@ class Interface : public Kernel::SessionRequestHandler {
...
+                  // Type-erase member function pointer by casting it down to the base class.
+                  static_cast<HandlerFnP<ServiceFrameworkBase>>(handler_callback), name} {}
+    };
+
+    /**
+     * Initializes the handler with no functions installed.
+     * @param max_sessions Maximum number of sessions that can be
+     * connected to this service at the same time.
+     */
+    ServiceFramework(const char* service_name, unsigned int max_sessions = DefaultMaxSessions)
```
Comment (lioncash) [inline src/core/hle/service/service.h]: ditto

File: src/core/hle/service/sm/srv.h [hunk 3]
```diff
@@ -4,21 +4,33 @@
...
 namespace Service {
 namespace SM {
 
 /// Interface to "srv:" service
-class SRV final : public Interface {
+class SRV final : public ServiceFramework<SRV> {
 public:
-    SRV();
-    ~SRV() override;
+    SRV(std::shared_ptr<ServiceManager> service_manager);
```
Comment (lioncash) [inline src/core/hle/service/sm/srv.h]: Should this be an explicit constructor?

File: src/core/hle/service/service.h (hunk 1 above)
Comment (yuriks) [inline src/core/hle/service/service.h]: Done.

File: src/core/hle/service/service.h (hunk 2 above)
Comment (yuriks) [inline src/core/hle/service/service.h]: Done.

**Pull request merged**